README
			  linuxconsole tools
			    Release 1.6.1
		 http://sf.net/projects/linuxconsole/


This package contains the following utilities and associated
documentation:
* inputattach - connects legacy serial devices to the input layer
* joystick utilities - calibrate and test joysticks and joypads

The evtest tool, which used to be part of this package, is now
maintained separately at http://cgit.freedesktop.org/evtest/ and
https://bugzilla.freedesktop.org/enter_bug.cgi?product=evtest


Joystick utilities
------------------

The following utilities are provided to calibrate and test joysticks:
* evdev-joystick - calibrate joystick devices (including dead zones
  and fuzz)
* ffcfstress, ffmvforce, fftest - test force-feedback devices
* ffset - set force-feedback device parameters
* jscal - calibrate joystick devices, reconfigure the axes and buttons
* jscal-store, jscal-restore - store and retrieve joystick device
  settings as configured using jscal
* jstest - test joystick devices

The typical scenario when configuring a new device is as follows:
1. Check the basic functions using jstest (number of buttons, axes,
   etc.).
2. Calibrate the joystick using jscal (this can be useful even if the
   device functions correctly, if only to configure the dead-zone at
   the centre).
3. Repeat steps 1 and 2 until the joystick is configured to the user's
   satisfaction.
4. Store the device's setup using jscal-store.


Installation
------------

Most of the utilities have no particular requirement beyond a libc and
the Linux input layer headers (normally part of your libc
installation). ffmvforce requires SDL (libsdl1.2-dev on Debian
derivatives).

To install the utilities and their manpages, run
	make
or
	make SYSTEMD_SUPPORT=1
and then as root
	make install
from the distribution's top directory. This will install everything
into /usr/local. Should you wish to install the tools elsewhere, you
can use the PREFIX variable:
	PREFIX=/usr make install
The Makefiles also recognise the DESTDIR variable which can be used to
install into a temporary directory (this is useful mainly for
distribution packagers).


Auto-loading inputattach and jscal-restore
------------------------------------------

udev can be used to start inputattach and jscal-restore when the
appropriate device is connected to the system (or detected when the
system is booting up).

The following rule restores the stored calibration and axis/button
mappings whenever a joystick device is connected:
	KERNEL=="js*", ACTION=="add", RUN+="/usr/bin/jscal-restore %E{DEVNAME}"
(change the path as appropriate). This rule needs /usr to be
available, which can cause issues on systems where udev is run from
the initramfs or where /usr is a separate partition. To work around
this, the Debian package uses a script provided by Debian's udev
package to wait for jscal-restore to become available. Should you wish
to reproduce this, you need to create a udev agent
(/lib/udev/jscal-restore) containing the following:

#!/bin/sh -e
# udev agent for jscal-restore

wait_for_file() {
	local file=$1
	local count=120
	while [ $count != 0 ]; do
		[ -e "$file" ] && return 0
		sleep 1
		count=$(($count - 1))
	done
	exit 1
}

(
	wait_for_file /usr/bin/jscal-restore
	exec /usr/bin/jscal-restore $1
) &

and change the udev rule to
	KERNEL=="js*", ACTION=="add", RUN+="/lib/udev/jscal-restore %E{DEVNAME}"

The following rule forces flaky Acrux gamepads to stay awake:
	KERNEL=="js*", ACTION=="add", ATTRS{name}=="ACRUX USB GAMEPAD 8116", RUN+="/bin/sleep 5000 < /dev/%k"
(This should only be necessary on kernels older than 2.6.39-rc1.)

The following rule configures a Mouse Systems mouse on the first
serial port:
	KERNEL=="ttyS0", ACTION=="add", RUN+="/usr/bin/inputattach --daemon -msc /dev/%k"
If you need to wait for /usr you can adapt the recipe above.

The following rules configures a Wacom W8001 devices on a Fujitsu
T2010[2] or on Lenovo X200-series laptops and tablets:
	SUBSYSTEM=="tty", KERNEL=="ttyS[0-9]*", ATTRS{id}=="FUJ02e5", ACTION=="add|change", RUN+="/usr/bin/inputattach --daemon --baud 19200 --w8001 /dev/%k"
	SUBSYSTEM=="tty", KERNEL=="ttyS[0-9]*", ATTRS{id}=="WACf00c", ACTION=="add|change", RUN+="/usr/bin/inputattach --daemon --baud 38400 --w8001 /dev/%k"

These rules are liable to cause conflicts with the X.org wacom
driver. For example, on some Debian-based systems the
/usr/share/X11/xorg.conf.d/50-wacom.conf configuration file needs to
be amended so that the "MatchProduct" line matching one of the
identifiers used above ("FUJ02e5" or "WACf00c", the latter being
matched by "WACf") no longer contains the matching identifier. The
appropriate X.org driver when using inputattach in this way is evdev.

A udev rule, especially shipped in a custom OS package is not
always applicable because it cannot handle the conflict presented
by the fact that not all computers are built the same and on one
computer, e.g ttyS1 can be attached to a specific touchscreen,
in another computer it can be attached to a different kind of
touchscreen or a different device or connected at all.
For this case, a wrapper script started from sysvinit, upstart, or
systemd is better suited. systemd needs special treatment inside
inputattach: a notification must be sent to it when inputattach
becomes "ready". To achieve this, the previously mentioned
"make SYSTEMD_SUPPORT=1" must be used to include this support.
Then a systemd unit file and the wrapper script can be added, similar
to the following:

----8<---- inputattach-wrapper.sh ----8<----
#!/bin/sh

SYSTEM=`detect_computer_type`
case $SYSTEM in
type1)
    inputattach --daemon -pm9k /dev/ttyS2
    ;;
type2)
    inputattach --daemon -mtouch /dev/ttyS2
    ;;
# and so on
*)
    # do nothing for other computers
    ;;
esac

# Exit without errors, so systemd won't try to
# restart our script for the "do nothing" case.
exit 0
----8<- end of inputattach-wrapper.sh-8<----

----8<---- inputattach.service    ----8<----
[Unit]
Description=inputattach helper service

[Service]
Type=forking
NotifyAccess=all
ExecStart=/usr/sbin/inputattach-wrapper.sh

[Install]
WantedBy=multi-user.target
----8<- end of inputattach.service----8<----

Configuring PS3 controller LEDs
-------------------------------

The js-set-enum-leds can be used with udev to ensure that the
controller LEDs on PS3 controllers match their identifiers as set by
Bluetooth host software. This ensures that if the controllers are
plugged in before the Bluetooth daemon starts, the LEDs still match
the identifier assigned once the daemon has initialised.

The following udev rule achieves this:

	# Set PS3 controller leds to the same value bluez will choose for it.
	# IMPORT{parent} used because $parent did not appear to be working; we
	# don't need the devpath of the js device for this so it is OK to overwrite.
	ACTION=="add", KERNEL=="js*", IMPORT{parent}="DEVPATH", ATTRS{name}=="Sony PLAYSTATION(R)3 Controller", RUN+="/lib/udev/js-set-enum-leds /sys/%E{DEVPATH}/device/leds/ $number"


Contributors
------------

The following people have contributed to this release:
* Johann Walles: jscal jitter tolerance.
* Dr. László Kaján: button and axis remapping.
* Sebastian Siewior: EVIOCGBIT correction.
* Krzysztof A. Sobiecki: improvements to the button and axis remapping
  functions.
* Claudio Nieder: Sahara Touch-iT213 support.
* Florian Fainelli: evtest fixes.
* Alexander Clouter: W8001 support and error-handling fixes.
* Roberto Neri: much discussion, and fixes and improvements to
  jscal-store/jscal-restore and the udev rules given above.
* Jean Delvare: inputattach improvements, TAOS support.
* Brian Murray and Kees Cook: Lenovo W8001 support.
* Raoul Bönish: fixes for the udev rules.
* John Sung: PenMount 3000, 6000 and 6250 support.
* Christian Gmeiner: TSC-40 support.
* Kyle Manna: various build fixes.
* Denis Gonsiorovsky: Genius EasyPen 3x4 tablet support.
* Jaromir Capik: 64-bit fixes and changes from ff-utils (based on a
  patch originally by Jean-Philippe Meure).
* Ralf Jung: fixes for long device names.
* Andrew Basterfield: fix for Spaceball 4000FLX support.
* bri: PS3 controller LED configuration
* Julian Squires: Wacom IV support.
* Jamie Lentin: mouseman4 mode.
* Böszörményi Zoltán: eGalaxTouch support, Hampshire support, systemd
  notification.
* Stephen Anthony: evdev-joystick (based on G25manage).
* Hans Verkuil: Pulse-Eight and RainShadow HDMI CEC dongle support.
* Jon Sangster: min/max axis control in evdev-joystick.


License
-------

The linuxconsole tools are
	Copyright © 1996-2005 Vojtech Pavlik
	Copyright © 2001 Oliver Hamann
	Copyright © 2001-2002 Johann Deneux
	Copyright © 2001 Arndt Schoenewald
	Copyright © 2008-2019 Stephen Kitt
	Copyright © 2016 Stephen Anthony


This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or (at
your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301,
USA.


The GNU General Public License version 2 is provided in the COPYING
file.


NEWS
Version 1.6.1
-------------

* inputattach supports the RainShadow HDMI CEC dongle (this requires
  kernel 4.12 or later; thanks to Hans Verkuil).

* The jscal store and restore tools use udevadm on the path, instead
  of hard-coding /sbin.

Version 1.6.0
-------------

* inputattach supports the Pulse-Eight HDMI CEC dongle (this requires
  kernel 4.8 or later; thanks to Hans Verkuil <hverkuil@xs4all.nl>).

* evdev-joystick can set the minimum and maximum axis values (thanks
  to Jon Sangster <jon@ertt.ca>).

* The udev rules are installed without executable permissions.

* The evdev-joystick manpage is installed along with the other
  manpages.

Version 1.5.1
-------------

* Correctly install 80-stelladaptor-joystick.rules.

Version 1.5.0
-------------

* Add Stephen Anthony's evdev-joystick tool, with his agreement.

Version 1.4.9
-------------

* inputattach supports a new mouseman4 mode for 4-button Logitech mice
  (thanks to Jamie Lentin <jm@lentin.co.uk>).

* inputattach supports EETI eGalaxTouch touchscreens (thanks to
  Böszörményi Zoltán <zboszor@pr.hu>).

* inputattach can provide a "ready" notification to systemd (thanks to
  Böszörményi Zoltán <zboszor@pr.hu>).

* js-set-enum-leds no longer uses hotplug.functions.

Version 1.4.8
-------------

* Set up PS3 controller LEDs correctly (thanks to bri
  <bri@abrij.org>).

* inputattach supports Wacom IV tablets on kernels 3.17 and greater
  (thanks to Julian Squires <julian@cipht.net>).

Version 1.4.7
-------------

* Correctly handle newer versions of the Spaceball firwmare (thanks to
  Andrew Basterfield <abasterfield@gmail.com>). This fixes Debian
  #298494.

Version 1.4.6
-------------

* The force-feedback test tools (ffmvforce, ffset, fftest) now handle
  device names longer than 64 characters (thanks to Ralf Jung
  <post@ralfj.de>). This fixes Debian #706744.

Version 1.4.5
-------------

* inputattach no longer loops forever in certain situations (thanks to
  Josh Bowman for the pointer).

* Improvements to the force-feedback tools, available elsewhere, have
  been merged (thanks to Jaromir Capik <jcapik@redhat.com>).

* ELO touchscreens apparently don't work with CRTSCTS, so it's no
  longer set (thanks to Thierry Bultel <thierry.bultel@wanadoo.fr>).
  This fixes Debian #699030.


Version 1.4.4
-------------

* jstest --select correctly resets the timeout, and no longer ends up
  using all the CPU.

* inputattach builds on kernels without TSC-40 support (thanks to Kyle
  Manna <kyle.manna@fuel7.com>).

* inputattach now supports Genius EasyPen 3x4 tablets on kernels with
  the appropriate module (thanks to Denis Gonsiorovsky
  <denis_g@outlook.com>).


Version 1.4.3
-------------

* inputattach supports PenMount 3000, 6000 and 6250 devices (thanks to
  John Sung <penmount.touch@gmail.com>) and TSC-40 devices (thanks to
  Christian Gmeiner <christian.gmainer@gmail.com>).


Version 1.4.2
-------------

* The udev rule matching w8001 devices has been split to handle
  devices using different baud rates, and the different baud rates are
  documented.

* inputattach no longer resets the line discipline before exiting when
  it discovers the device is already managed by another inputattach
  instance.

* jscal-store creates its target directory if necessary.


Version 1.4.1
-------------

* inputattach correctly handles non-retry errors other than EINTR
  (thanks to Alexander Clouter <alex@digriz.org.uk> for the patch).


Version 1.4
-----------

* inputattach no longer aborts when it receives a recoverable signal
  (thanks to Kees Cook <kees@ubuntu.com> for the patch).


Version 1.4rc1
--------------

The following user-visible changes have been made since the last
commonly used snapshot (20051019):

* inputattach has been cleaned up, and now supports the following new
  devices:
  - Sahara Touch-iT213
  - PS/2 multiplexers on TQM85xx boards
  - Zhen Hua PCM-4CH RC transmitters
  - TAOS evaluation modules
  - Wacom W8001
  - Microtouch, Touchright, Touchwindow, Penmount and Fujitsu
    touchscreens

* jscal tolerates greater jitter (so joysticks which don't send stable
  values can still be calibrated).

* jscal supports axis and button remapping.

* Various USB devices which caused jscal and/or jstest to segfault are
  now handled correctly.

* jscal-store and jscal-restore allow joystick calibration and
  remapping settings to be stored and restored.

* Documentation is provided, with manpages for all the provided
  utilities and sample udev rules.

In addition, the 20051019 snapshot included the following user-visible
changes since version 1.2.15:

* jstest includes axis and button names.

* New utilities:
  - inputattach connects devices to the input layer, allowing them to
    be used even if they are not directly supported;
  - force-feedback utilities;
  - evtest event-layer monitoring utility.


utils/serio-ids.h
#ifndef _SERIO_IDS_H
#define _SERIO_IDS_H

#ifndef SERIO_RS232
# define SERIO_RS232	0x02
#endif

/*
 * Serio types
 */
#ifndef SERIO_UNKNOWN
# define SERIO_UNKNOWN		0x00
#endif
#ifndef SERIO_MSC
# define SERIO_MSC		0x01
#endif
#ifndef SERIO_SUN
# define SERIO_SUN		0x02
#endif
#ifndef SERIO_MS
# define SERIO_MS		0x03
#endif
#ifndef SERIO_MP
# define SERIO_MP		0x04
#endif
#ifndef SERIO_MZ
# define SERIO_MZ		0x05
#endif
#ifndef SERIO_MZP
# define SERIO_MZP		0x06
#endif
#ifndef SERIO_MZPP
# define SERIO_MZPP		0x07
#endif
#ifndef SERIO_VSXXXAA
# define SERIO_VSXXXAA		0x08
#endif
#ifndef SERIO_SUNKBD
# define SERIO_SUNKBD		0x10
#endif
#ifndef SERIO_WARRIOR
# define SERIO_WARRIOR		0x18
#endif
#ifndef SERIO_SPACEORB
# define SERIO_SPACEORB		0x19
#endif
#ifndef SERIO_MAGELLAN
# define SERIO_MAGELLAN		0x1a
#endif
#ifndef SERIO_SPACEBALL
# define SERIO_SPACEBALL	0x1b
#endif
#ifndef SERIO_GUNZE
# define SERIO_GUNZE		0x1c
#endif
#ifndef SERIO_IFORCE
# define SERIO_IFORCE		0x1d
#endif
#ifndef SERIO_STINGER
# define SERIO_STINGER		0x1e
#endif
#ifndef SERIO_NEWTON
# define SERIO_NEWTON		0x1f
#endif
#ifndef SERIO_STOWAWAY
# define SERIO_STOWAWAY		0x20
#endif
#ifndef SERIO_H3600
# define SERIO_H3600		0x21
#endif
#ifndef SERIO_PS2SER
# define SERIO_PS2SER		0x22
#endif
#ifndef SERIO_TWIDKBD
# define SERIO_TWIDKBD		0x23
#endif
#ifndef SERIO_TWIDJOY
# define SERIO_TWIDJOY		0x24
#endif
#ifndef SERIO_HIL
# define SERIO_HIL		0x25
#endif
#ifndef SERIO_SNES232
# define SERIO_SNES232		0x26
#endif
#ifndef SERIO_SEMTECH
# define SERIO_SEMTECH		0x27
#endif
#ifndef SERIO_LKKBD
# define SERIO_LKKBD		0x28
#endif
#ifndef SERIO_ELO
# define SERIO_ELO		0x29
#endif
#ifndef SERIO_MICROTOUCH
# define SERIO_MICROTOUCH	0x30
#endif
#ifndef SERIO_PENMOUNT
# define SERIO_PENMOUNT		0x31
#endif
#ifndef SERIO_TOUCHRIGHT
# define SERIO_TOUCHRIGHT	0x32
#endif
#ifndef SERIO_TOUCHWIN
# define SERIO_TOUCHWIN		0x33
#endif
#ifndef SERIO_TAOSEVM
# define SERIO_TAOSEVM		0x34
#endif
#ifndef SERIO_FUJITSU
# define SERIO_FUJITSU		0x35
#endif
#ifndef SERIO_ZHENHUA
# define SERIO_ZHENHUA		0x36
#endif
#ifndef SERIO_INEXIO
# define SERIO_INEXIO		0x37
#endif
#ifndef SERIO_TOUCHIT213
# define SERIO_TOUCHIT213	0x38
#endif
#ifndef SERIO_W8001
# define SERIO_W8001		0x39
#endif
#ifndef SERIO_PSMULT
# define SERIO_PS2MULT		0x3c
#endif
#ifndef SERIO_WACOM_IV
# define SERIO_WACOM_IV		0x3d
#endif
#ifndef SERIO_WACOM_V
# define SERIO_WACOM_V		0x3e
#endif
#ifndef SERIO_EASYPEN
# define SERIO_EASYPEN		0x3f
#endif
#ifndef SERIO_PULSE8_CEC
# define SERIO_PULSE8_CEC	0x40
#endif
#ifndef SERIO_RAINSHADOW_CEC
# define SERIO_RAINSHADOW_CEC	0x41
#endif

#endif


utils/evdev-joystick.c
/**
  This program is based on G25manage, located at:
    https://github.com/VDrift/vdrift/tree/master/tools/G25manage

  This code is released under the GPLv2, and modified from the
  original by Stephen Anthony (stephena@users.sf.net).
*/

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <getopt.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <stdio.h>
#include <errno.h>
#include <stdint.h>
#include <string.h>
#include <dirent.h>

#include <sys/ioctl.h>
#include <sys/types.h>
#include <asm/types.h>
#include <fcntl.h>

#include <linux/input.h>

/* this macro is used to tell if "bit" is set in "array"
 * it selects a byte from the array, and does a boolean AND 
 * operation with a byte that only has the relevant bit set. 
 * eg. to check for the 12th bit, we do (array[1] & 1<<4)
 */
#define test_bit(bit, array)    (array[bit/8] & (1<<(bit%8)))

// The default location for evdev devices in Linux
#define EVDEV_DIR "/dev/input/by-id/"

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void help()
{
  printf("%s","Usage:\n\n"
    "  --help, --h              The message you're now reading\n"
    "  --listdevs, --l          List all joystick devices found\n"
    "  --showcal, --s [path]    Show current calibration for joystick device\n"
    "  --evdev, --e [path]      Set the joystick device to modify\n"
    "  --minimum, --m [val]     Change minimum for current joystick\n"
    "  --maximum, --M [val]     Change maximum for current joystick\n"
    "  --deadzone, --d [val]    Change deadzone for current joystick\n"
    "  --fuzz, --f [val]        Change fuzz for current joystick\n"
    "  --axis, --a [val]        The axis to modify for current joystick (by default, all axes)\n"
    "\n"
    "To see calibration information: \n"
    "  evdev-joystick [ --s /path/to/event/device/file ]\n"
    "\n"
    "To set the deadzone values:\n"
    "  evdev-joystick [ --e /path/to/event/device/file --d deadzone_value [ --a axis_index ] ]\n"
    "\n"
    "Example:\n"
    "\n"
    "I want to see the calibration values of my event managed joystick:\n"
    "  evdev-joystick --s /dev/input/event6\n"
    "\n"
    "Supported Absolute axes:\n"
    "  Absolute axis 0x00 (0) (X Axis) (min: 0, max: 16383, flatness: 1023 (=6.24%), fuzz: 63)\n"
    "  Absolute axis 0x01 (1) (Y Axis) (min: 0, max: 255, flatness: 15 (=5.88%), fuzz: 0)\n"
    "  Absolute axis 0x02 (2) (Z Axis) (min: 0, max: 255, flatness: 15 (=5.88%), fuzz: 0)\n"
    "  Absolute axis 0x05 (5) (Z Rate Axis) (min: 0, max: 255, flatness: 15 (=5.88%), fuzz: 0)\n"
    "  Absolute axis 0x10 (16) (Hat zero, x axis) (min: -1, max: 1, flatness: 0 (=0.00%), fuzz: 0)\n"
    "  Absolute axis 0x11 (17) (Hat zero, y axis) (min: -1, max: 1, flatness: 0 (=0.00%), fuzz: 0)\n"
    "\n"
    "I want to get rid of the deadzone on all axes on my joystick:\n"
    "  evdev-joystick --e /dev/input/event6 --d 0\n"
    "\n");
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void listDevices()
{
  DIR* dirp = opendir(EVDEV_DIR);
  struct dirent* dp;

  if(dirp == NULL)
    return;

  // Loop over dir entries using readdir
  int len = strlen("event-joystick");
  while((dp = readdir(dirp)) != NULL)
  {
    // Only select names that end in 'event-joystick'
    int devlen = strlen(dp->d_name);
    if(devlen >= len)
    {
      const char* start = dp->d_name + devlen - len;
      if(strncmp(start, "event-joystick", len) == 0)
        printf("%s%s\n", EVDEV_DIR, dp->d_name);
    }
  }
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void printAxisType(int i)
{
  switch(i)
  {
    case ABS_X :        printf(" (X Axis) ");             break;
    case ABS_Y :        printf(" (Y Axis) ");             break;
    case ABS_Z :        printf(" (Z Axis) ");             break;
    case ABS_RX :       printf(" (X Rate Axis) ");        break;
    case ABS_RY :       printf(" (Y Rate Axis) ");        break;
    case ABS_RZ :       printf(" (Z Rate Axis) ");        break;
    case ABS_THROTTLE : printf(" (Throttle) ");           break;
    case ABS_RUDDER :   printf(" (Rudder) ");             break;
    case ABS_WHEEL :    printf(" (Wheel) ");              break;
    case ABS_GAS :      printf(" (Accelerator) ");        break;
    case ABS_BRAKE :    printf(" (Brake) ");              break;
    case ABS_HAT0X :    printf(" (Hat zero, x axis) ");   break;
    case ABS_HAT0Y :    printf(" (Hat zero, y axis) ");   break;
    case ABS_HAT1X :    printf(" (Hat one, x axis) ");    break;
    case ABS_HAT1Y :    printf(" (Hat one, y axis) ");    break;
    case ABS_HAT2X :    printf(" (Hat two, x axis) ");    break;
    case ABS_HAT2Y :    printf(" (Hat two, y axis) ");    break;
    case ABS_HAT3X :    printf(" (Hat three, x axis) ");  break;
    case ABS_HAT3Y :    printf(" (Hat three, y axis) ");  break;
    case ABS_PRESSURE : printf(" (Pressure) ");           break;
    case ABS_DISTANCE : printf(" (Distance) ");           break;
    case ABS_TILT_X :   printf(" (Tilt, X axis) ");       break;
    case ABS_TILT_Y :   printf(" (Tilt, Y axis) ");       break;
    case ABS_MISC :     printf(" (Miscellaneous) ");      break;
    default:            printf(" (Unknown absolute feature) ");
  }
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
int showCalibration(const char* evdev)
{
  int fd = -1, axisindex;
  uint8_t abs_bitmask[ABS_MAX/8 + 1];
  float percent_deadzone;
  struct input_absinfo abs_features;

  if((fd = open(evdev, O_RDONLY)) < 0)
  {
    perror("evdev open");
    return 1;
  }

  memset(abs_bitmask, 0, sizeof(abs_bitmask));
  if (ioctl(fd, EVIOCGBIT(EV_ABS, sizeof(abs_bitmask)), abs_bitmask) < 0)
    perror("evdev ioctl");

  printf("Supported Absolute axes:\n");

  for(axisindex = 0; axisindex < ABS_MAX; ++axisindex)
  {
    if(test_bit(axisindex, abs_bitmask))
    {
      // This means that the bit is set in the axes list
      printf("  Absolute axis 0x%02x (%d)", axisindex, axisindex);
      printAxisType(axisindex);

      if(ioctl(fd, EVIOCGABS(axisindex), &abs_features))
        perror("evdev EVIOCGABS ioctl");

      percent_deadzone = (float)abs_features.flat * 100 / (float)abs_features.maximum;
      printf("(value: %d, min: %d, max: %d, flatness: %d (=%.2f%%), fuzz: %d)\n",
        abs_features.value, abs_features.minimum, abs_features.maximum,
        abs_features.flat, percent_deadzone, abs_features.fuzz);
    }
  }

  close(fd);
  return 0;
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
int setAxisInfo(const char* evdev, int axisindex,
                       __s32 minvalue, __s32 maxvalue,
                       __s32 deadzonevalue, __s32 fuzzvalue)
{
  int fd = -1;
  uint8_t abs_bitmask[ABS_MAX/8 + 1];
  float percent_deadzone;
  struct input_absinfo abs_features;

  if ((fd = open(evdev, O_RDONLY)) < 0)
  {
    perror("evdev open");
    return 1;
  }

  memset(abs_bitmask, 0, sizeof(abs_bitmask));
  if(ioctl(fd, EVIOCGBIT(EV_ABS, sizeof(abs_bitmask)), abs_bitmask) < 0)
    perror("evdev ioctl");

  int axis_first = 0, axis_last = ABS_MAX;
  if(axisindex >= 0 && axisindex < ABS_MAX)
  {
    axis_first = axisindex;
    axis_last = axisindex + 1;
  }

  for(axisindex = axis_first; axisindex < axis_last; ++axisindex)
  {
    if(test_bit(axisindex, abs_bitmask))
    {
      /* this means that the bit is set in the axes list */
      printf("  Absolute axis 0x%02x (%d)", axisindex, axisindex);
      printAxisType(axisindex);

      if(ioctl(fd, EVIOCGABS(axisindex), &abs_features))
      {
        perror("evdev EVIOCGABS ioctl");
        return 1;
      }

      if(minvalue != INT_MIN)
      {
        printf("Setting min value to : %d\n", minvalue);
        abs_features.minimum = minvalue;
      }

      if(maxvalue != INT_MIN)
      {
        printf("Setting max value to : %d\n", maxvalue);
        abs_features.maximum = maxvalue;
      }

      if(deadzonevalue != INT_MIN)
      {
        if(deadzonevalue < abs_features.minimum ||
           deadzonevalue > abs_features.maximum )
        {
          printf("Deadzone value must be between %d and %d for this axis, "
                 "value requested : %d\n",
            abs_features.minimum, abs_features.maximum, deadzonevalue);
        }

        printf("Setting deadzone value to : %d\n", deadzonevalue);
        abs_features.flat = deadzonevalue;
      }

      if(fuzzvalue != INT_MIN)
      {
        if(fuzzvalue < abs_features.minimum ||
           fuzzvalue > abs_features.maximum )
        {
          printf("Fuzz value must be between %d and %d for this axis, "
                 "value requested : %d\n",
            abs_features.minimum, abs_features.maximum, fuzzvalue);
        }

        printf("Setting fuzz value to : %d\n", fuzzvalue);
        abs_features.fuzz = fuzzvalue;
      }

      if(ioctl(fd, EVIOCSABS(axisindex), &abs_features))
      {
        perror("evdev EVIOCSABS ioctl");
        return 1;
      }
      if(ioctl(fd, EVIOCGABS(axisindex), &abs_features))
      {
        perror("evdev EVIOCGABS ioctl");
        return 1;
      }
      percent_deadzone = (float)abs_features.flat * 100 / (float)abs_features.maximum;
      printf("    (value: %d, min: %d, max: %d, flatness: %d (=%.2f%%), fuzz: %d)\n",
        abs_features.value, abs_features.minimum, abs_features.maximum,
        abs_features.flat, percent_deadzone, abs_features.fuzz);
    }
  }

  close(fd);
  return 0;
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
int main(int argc, char* argv[])
{
  char* evdevice = NULL;
  int c, axisindex = -1;
  __s32 min = INT_MIN, max = INT_MIN, flat = INT_MIN, fuzz = INT_MIN;

  // Show help by default
  if(argc == 1)
  {
    help();
    exit(0);
  }

  while(1)
  {
    static struct option long_options[] =
    {
      { "help",     no_argument,       0, 'h' },
      { "listdevs", no_argument,       0, 'l' },
      { "showcal",  required_argument, 0, 's' },
      { "evdev",    required_argument, 0, 'e' },
      { "minimum",  required_argument, 0, 'm' },
      { "maximum",  required_argument, 0, 'M' },
      { "deadzone", required_argument, 0, 'd' },
      { "fuzz",     required_argument, 0, 'f' },
      { "axis",     required_argument, 0, 'a' },
      { 0, 0, 0, 0 }
    };
    // getopt_long stores the option index here
    int option_index = 0;

    c = getopt_long(argc, argv, "h:l:s:e:d:m:M:f:a:", long_options, &option_index);

    // Detect the end of the options
    if(c == -1)
      break;

    switch(c)
    {
      case 0:
        // If this option set a flag, do nothing else now.
        if(long_options[option_index].flag != 0)
          break;
        printf("option %s", long_options[option_index].name);
        if(optarg)
          printf(" with arg %s", optarg);
        printf("\n");
        break;

      case 'h':
        help();
        break;

      case 'l':
        listDevices();
        break;

      case 's':
        evdevice = optarg;
        showCalibration(evdevice);
        break;

      case 'e':
        evdevice = optarg;
        printf("Event device file: %s\n", evdevice);
        break;

      case 'd':
        flat = atoi(optarg);
        printf("New dead zone value: %d\n", flat);
        break;

      case 'm':
        min = atoi(optarg);
        printf("New min value: %d\n", min);
        break;

      case 'M':
        max = atoi(optarg);
        printf("New max value: %d\n", max);
        break;

      case 'f':
        fuzz = atoi(optarg);
        printf("New fuzz value: %d\n", fuzz);
        break;

      case 'a':
        axisindex = atoi(optarg);
        printf("Axis index to deal with: %d\n", axisindex);
        break;

      case '?':
        // getopt_long already printed an error message.
        break;

      default:
        abort();
    }
  }

  // Print any remaining command line arguments (not options).
  if(optind < argc)
  {
    printf("non-option ARGV-elements: ");
    while(optind < argc)
      printf("%s ", argv[optind++]);
    putchar('\n');
  }

  if(min != INT_MIN || max != INT_MIN || flat != INT_MIN || fuzz != INT_MIN)
  {
    if(evdevice == NULL)
    {
      printf( "You must specify the event device for your joystick\n" );
      exit(1);
    }
    else
    {
      if(axisindex == -1)
      {
        if(min != INT_MIN)
          printf( "Trying to set all axes minimum to: %d\n", min);
        if(max != INT_MIN)
          printf( "Trying to set all axes maximum to: %d\n", max);
        if(flat != INT_MIN)
          printf( "Trying to set all axes deadzone to: %d\n", flat);
        if(fuzz != INT_MIN)
          printf( "Trying to set all axes fuzz to: %d\n", fuzz);
      }
      else
      {
        if(min != INT_MIN)
          printf( "Trying to set axis %d minimum to: %d\n", axisindex, min);
        if(max != INT_MIN)
          printf( "Trying to set axis %d maximum to: %d\n", axisindex, max);
        if(flat != INT_MIN)
          printf( "Trying to set axis %d deadzone to: %d\n", axisindex, flat);
        if(fuzz != INT_MIN)
          printf( "Trying to set axis %d fuzz to: %d\n", axisindex, fuzz);
      }

      setAxisInfo(evdevice, axisindex, min, max, flat, fuzz);
    }
  }

  exit(0);
}
